README.md
![StreamDude Logo](assets/logos/streamdude-logo-128x128.png)

# StreamDude

[![Go](https://github.com/GwynethLlewelyn/StreamDude/actions/workflows/go.yml/badge.svg)](https://github.com/GwynethLlewelyn/StreamDude/actions/workflows/go.yml) [![Codacy Security Scan](https://github.com/GwynethLlewelyn/StreamDude/actions/workflows/codacy.yml/badge.svg)](https://github.com/GwynethLlewelyn/StreamDude/actions/workflows/codacy.yml)

A work in progress simple playlist manager to remotely send files to a streaming server.

Licensed under a [MIT License](https://gwyneth-llewelyn.mit-license.org/).

## Purpose

There are a million gorgeous tools out there which allow you to manage your local playlist and send it to a streaming server; there are million more — such as OBS — which, while not _stricly_ designed to do just that, can very easily be adapted for that purpose. And there are a *b*illion that even allow you to stream (using so-called progressive download) files from a webpage to your friends?

But what if your music is not stored locally, or you wish to avoid the inevitable delays (latency and/or jitter) for sending the files from your computer to a remote server, and stream from _there_? Think of an Internet radio station, managed by several parties, with a common repository of music (or video!), all stored and managed remotely. Oh, of course: and for free.

Well, you're out of luck.

The oldest tool I know that does just that is Apple's own Darwin Streaming Server — which is even free and open-source — but it's overkill if you already have an external streaming server that you connect with.

There are a few complex ones out there, some even for free, but installing them is a pain.

Thus, StreamDude.

## No, really?

Well, no, I'm lying.

Actually what I _really_ need **right now** is a simple way to get scripts from inside the [Second Life®](https://secondlife.com) and [OpenSimulator](http://opensimulator.org) to be able to remotely select what streams to play. I could do it with a single Python command line argument (no need to run or compile anything or write any code whatsoever). But what would be the fun in that?

Instead, I'm sort of joining three separate projects into one.

## Environment variables

-   `LAL_MASTER_KEY` - because it's too dangerous to keep it in code and/or files
-   `STREAMER_URL` - another way to override the streamer URL; may be useful in scripts

Also, StreamDude attempts to comply with the informal `CLICOLOR_FORCE` and `NO_COLOR` conventions. See https://bixense.com/clicolors/ and https://no-color.org/.

## Compile & launch

You can get [here the full API for Postman](extras/StreamDude.postman_collection.json).

Currently, the only streaming server supported is [lal (Live And Live)](https://github.com/q191201771/lal).

1. Make sure you have the streaming server running first!
2. `go install github.com/GwynethLLewelyn/StreamDude@latest` or, if you prefer, `git clone https://github.com/GwynethLLewelyn/StreamDude`.
3. If you cloned the repo, then run `go build` (and possibly with `go install` you'll get the compiled binary under `~/go/bin`, which, hopefully, is part of your `$PATH`)
4. `LAL_MASTER_KEY=blahblehblih ./StreamDude -d` (if you wish debugging to console, or redirect it to a log file)
5. `/usr/bin/curl --header "Content-Type: application/json" --header "Accept: application/json" --request GET    http://127.0.0.1:3554/ping` — should give `{"message":"pong back to 127.0.0.1","status":"ok"}`
6. `/usr/bin/curl --header "Content-Type: application/json" --header "Accept: application/json" --request POST   --data '{ "objectPIN": "0000" }' http://127.0.0.1:3554/api/auth` — should give you an authentication token, e.g. `ZmFrZXRva2Vu`
7. `/usr/bin/curl --header "Content-Type: application/json" --header "Accept: application/json" --request POST   --data '{ "token": "ZmFrZXRva2Vu", "filename": "/path/to/video.mp4"  }' http://127.0.0.1:3554/api/play` — should launch ffmpeg and send `video.mp4` to be streamed

**Note:** `objectPIN` and `token` are not really, really being enforced — there is no database/KV store backend yet, but as soon as there is one, I've put the validation code in place, so you should fill in those fields.

Also note that there are further fields for Second Life®/OpenSimulator, all of which are being ignored right now.

## Backoffice

Currently not implemented. There is just a home page and a Ping test. The remaining menu areas are not really working (they will become forms for testing purposes).

The home page, properly speaking, will become an instance of [MusicFolderPlayer](https://github.com/ltguillaume/music-folder-player/) (the inspiration for this project).

## Nginx conf sample

### Assumptions

1. You're running Unix-like environment (WSL2 will possibly work, too);
2. `nginx` is pointing to `http(s)://my.streaming.server` under `/var/www/my.streaming.server`;
3. **StreamDude** is installed under `/StreamDude` (i.e. `/templates` and `/assets` are there)
4. It runs as a service under `localhost:3445` (default)

Then you will need something like this:

```nginx
location ~* /StreamDude/assets {
	rewrite ^/StreamDude/assets/(.*)$ /$1 break;
	root /var/www/my.streaming.server/StreamDude/assets;
	expires max;
	add_header Cache-Control public;
	fastcgi_hide_header Set-Cookie;
	try_files $uri =404;
}

location ~* /StreamDude/templates {
	rewrite ^/StreamDude/templates/(.*)$ /$1 break;
	root /var/www/my.streaming.server/StreamDude/templates;
	expires max;
	add_header Cache-Control public;
	fastcgi_hide_header Set-Cookie;
	try_files $uri =404;
}

location /StreamDude/ {
	proxy_pass_request_headers on;
	proxy_pass_request_body on;
	proxy_set_header X-Forwarded-For   $remote_addr;
	proxy_set_header X-Forwarded-Proto $scheme;
	proxy_set_header Host              $host;
	proxy_set_header X-Real-IP         $remote_addr;
	proxy_pass_header CF-Connecting-IP;
	proxy_pass_header CF-IPCountry;
	proxy_pass_header Set-Cookie;
	proxy_buffering off;
	proxy_ssl_server_name on;
	proxy_read_timeout 5m;
	proxy_set_header Access-Control-Allow-Credentials true;
	proxy_set_header Content-Encoding gzip;
	proxy_pass http://127.0.0.1:3554;
}

location ~* ^.+\.(xml|ogg|ogv|svg|svgz|eot|otf|woff|mp4|ttf|css|rss|atom|js|jpg|jpeg|gif|png|ico|zip|tgz|gz|rar|bz2|doc|xls|exe|ppt|tar|mid|midi|wav|bmp|rtf|lsl|lua)$ {
		access_log off; log_not_found off; expires max;
		add_header Cache-Control public;
		fastcgi_hide_header Set-Cookie;
}
```

and launch StreamDude (in debug mode) with:

```bash
$ ./StreamDude -d -r rtsp://127.0.0.1:5544/ -u /StreamDude -x my.streaming.server
```

Add `-P ":443"` if your front-end server is running HTTPS.

If you're launching StreamDude directly from the root of your virtual host (i.e. no `/StreamDude` subfolder), then you might need to add a trailing slash on `proxy_pass http://127.0.0.1:3554/;`. Getting the slashes to match properly is always messy.

## Launching from `systemd`

If you're running a Unix version supporting `systemd`, you can grab a [sample unit service file](extras/StreamDude.service.sample) to adapt to your needs. StreamDude complies with the [`sd_notify`](https://www.man7.org/linux/man-pages/man3/sd_notify.3.html) specifications and tries to play nicely with `systemd`.

Coloured `journald` logs are yet to be implemented, but at least you can get them using `journalctl -u StreamDude -f`.

## Release notes

See [CHANGELOG.md](CHANGELOG.md).


CHANGELOG.md
# Release notes for StreamDude

(most recent first)

## v0.0.2 (2023/08/08)

-   Added functional web backend (some things are missing)
-   Code refactoring: simplifying communication with Gin
-   Most functionality is equivalently available via HTML/JSON/XML or plain text
-   Code refactoring: moving UI-related functions to ui.go
-   To-do: check deeper on streaming errors that are passed back but silently ignored
-   To-do: create an uniform response generation for HTML/JSON/XML or plain text, to avoid repeating all that code over and over again

## v0.0.1 (2023/08/04)

-   First sort-of-working release


ui.go
// Web user interface
//
// © 2023 by Gwyneth Llewelyn. All rights reserved.
// Licensed under a MIT License (see https://gwyneth-llewelyn.mit-license.org/).
package main

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/gin-gonic/gin/binding"
)


/*
 *  Some handlers called directly here
 */

// Homepage is the front-end's first page. It might get some authentication at sme point.
func homepage(c *gin.Context) {
	responseContent := getContentType(c)

	// Default message for those who do NOT use application/html!
	homepageMessage := "It works. You should see it in HTML instead, it's so much nicer!"
	logme.Debugf("homepage: response Content-Type: %q; Request method: %q\n", responseContent, c.Request.Method)

//	switch getContentType(c) {
	switch responseContent {
		case binding.MIMEJSON:
			c.JSON(http.StatusOK, gin.H{"status": "ok", "message": homepageMessage})
		case binding.MIMEHTML, binding.MIMEPOSTForm, binding.MIMEMultipartPOSTForm:
			c.HTML(http.StatusOK, "home.tpl", environment(c, gin.H{
				"Title"			: "Welcome!",
				"description"	: "StreamDude demo homepage",
				"Text"			: "This is StreamDude — nothing will work on the menus, except Ping.",
			}))
		case binding.MIMEXML, "application/soap+xml", binding.MIMEXML2:	// we'll probably ignore soap/xml
			c.XML(http.StatusOK, gin.H{"status": "ok", "message": homepageMessage})
		case binding.MIMEPlain:
			fallthrough
		default:
			c.String(http.StatusOK, homepageMessage)
	}
}

// uiPing is the all-purpose ping testing function. Works with HTML too.
func uiPing(c *gin.Context) {
	responseContent := getContentType(c)

	// this will work even behind Cloudflare (gwyneth 20230804)
	payload := "pong back to " + c.ClientIP()
	logme.Debugf("Ping request (%s) from %q received; replying with Content-Type: %q\n", c.Request.Method, payload, responseContent)

	switch responseContent {
		case binding.MIMEJSON:
			c.JSON(http.StatusOK, gin.H{"status": "ok", "message": payload})
		case binding.MIMEHTML, binding.MIMEPOSTForm, binding.MIMEMultipartPOSTForm:
			// if we're behind Cloudflare, we can get a cute emoji flag
			// telling us which country this ping came from! (gwyneth 20230804)
			cfIPCountry := c.GetHeader("CF-IPCountry")
			if cfIPCountry != "" {			// this will usually be set by Cloudflare, too
				payload += " " + getFlag(cfIPCountry)
			}
			c.HTML(http.StatusOK, "generic.tpl", environment(c, gin.H{
				"Title"			: "Ping results",
				"description"	: http.StatusText(http.StatusOK) + " " + payload,
				"Text"			: payload,
			}))
		case binding.MIMEXML, "application/soap+xml", binding.MIMEXML2:
			c.XML(http.StatusOK, gin.H{"status": "ok", "message": payload})
		case binding.MIMEPlain:
			fallthrough
		default:
			c.String(http.StatusOK, payload)
	}
}

// Displays credits for the software. Only configured for HTML outpit.
func uiCredits(c *gin.Context) {
	c.HTML(http.StatusOK, "generic.tpl", environment(c, gin.H{
		"Title"			: "Credits",
		"description"	: "Credits",
		"Text"			: "One day, we will credit here everybody."		}))
}

assets/js/random-people.js
// this checks for the URL to see if it has an UUID as a parameter
function getURLParameter(name) {
	return decodeURIComponent((new RegExp('[?|&]' + name + '=' + '([^&;]+?)(&|#|;|$)').exec(location.search) || [null, ''])[1].replace(/\+/g, '%20')) || null;
}


// simple random name generator
function fetchNames(nameType) {
  let names = [];

  switch (nameType) {
	case 'female':
	  names = ['Berthefried', 'Tatiana', 'Hildeburg', 'Lily', 'Daisy'];
	  break;
	case 'male':
	  names = ['Bilbo', 'Frodo', 'Theodulph', 'Lotho'];
	  break;
	case 'surnames':
	  names = ['Baggins', 'Lightfoot', 'Boulderhill', 'Brockhouse', 'Boffin'];
	  break;
  }

  return { data: names };
}

function pickRandom(list) {
  return list[Math.floor(Math.random() * list.length)];
}

function generateName(gender) {
  // Fetch the names
  const firstNames = fetchNames(gender || pickRandom(['male', 'female']));
  const lastNames = fetchNames('surnames');

  // Pick a random name from each list
  const firstName = pickRandom(firstNames.data);
  const lastName = pickRandom(lastNames.data);

  // Use a template literal to format the full name
  return `${firstName} ${lastName}`;
}

streamdude.go
// StreamDude
//
// Main is here.
//
// Environment variables used when launching:
//
// `LAL_MASTER_KEY` - because it's too dangerous to keep it in code and/or files
// `STREAMER_URL` - another way to override the streamer URL; may be useful in scripts
//
// © 2023 by Gwyneth Llewelyn. All rights reserved.
// Licensed under a MIT License (see https://gwyneth-llewelyn.mit-license.org/).
//
package main

import (
	//	"log"
	"fmt"
	"net/http"
	"os"
	"os/signal"
	"path"
	"path/filepath"
	"syscall"

	"github.com/coreos/go-systemd/v22/daemon"

	"github.com/gin-gonic/gin"
//	"github.com/gin-gonic/gin/binding"
	"github.com/go-playground/validator/v10"

	//	"github.com/google/martian/log"
	flag "github.com/karrick/golf" // flag replacement library
	"github.com/mattn/go-isatty"
	"github.com/sirupsen/logrus"
)

// Globals

var (
	help bool					// if set, show usage
	ffmpegPath string			// path to ffmpeg executable
	ginMode *string				// ginMode is `debug` for development, `release` for production.
	host string					// this host — where StreamDude is running.
	serverPort string			// port where StreamDude server is listening
	frontEnd string				// FrontEnd is usually nginx but will probably be ignored later on.
	externalPort string			// external port if using a reverse proxy
	externalHost string			// external hostname if using a reverse proxy
	templatePath string			// where templates are held
	pathToStaticFiles string	// where static assets are stored
	workingDirectory string		// workingDirectory is the result of os.Getwd() or "." if that fails.
	urlPathPrefix string		// URL path prefix
	lslSignaturePIN string		// what we send from LSL
	debug bool					// set to debug level
	activeSystemd bool	= true	// if set, systemd is available (checked on start)

	// use a single instance of Validate, it caches struct info
	validate *validator.Validate

	// Global logger using logrus.
	logme = logrus.New()

	// Stuff for the lal streaming server
	streamerURL string		// RTSP streaming URL for lal
	lalMasterKey string		// too dangerous to show, put into LAL_MASTER_KEY environment
)

/*
 *  Ye Olde Maine Starts Here!
 *  Here Be Dragons
 *  And Unicorns
 */
func main() {
	// talk to systemd, inform that we're reloading
	b, err := daemon.SdNotify(false, daemon.SdNotifyReloading)
	// (false, nil) - notification not supported (i.e. NOTIFY_SOCKET is unset)
	// (false, err) - notification supported, but failure happened (e.g. error connecting to NOTIFY_SOCKET or while sending data)
	// (true, nil) - notification supported, data has been sentif
	switch {
		case !b && err == nil:
			// the logging system is not available, either, so we just print out
			fmt.Println("[WARN] systemd not available")
			activeSystemd = false
		case !b && err != nil:
			fmt.Println("[WARN] systemd answered with error:", err)
		case b && err == nil:
			fmt.Println("[INFO] systemd was succesfully notified that we're starting")
		default:
			fmt.Println("[WARN] unknown/confused systemd status, ignoring")
	}

	// Note: postponing some of the error logging until we know if the terminal is set and
	// handle colours etc. The above are "emergency" messages that need to be sent out before
	// the logging system is configured (namely, the first action should be) (gwyneth 20230807)

	// get the hostname, which is just used once, though (can be overriden with command-line arguments)
	hostname, err := os.Hostname()
	if err != nil {
		// who cares what the error was... in any case, we don't have the logging system yet:
		fmt.Printf("[WARN] system hostname not found (%s), using localhost instead\n", err)
		hostname = "localhost"
	}

	// Extract things from command line
	flag.BoolVarP(&help,			'h', "help",			false, 			"show command usage")
	flag.StringVarP(&ffmpegPath,	'm', "ffmpeg",			"/usr/local/bin/ffmpeg", "path to ffmpeg executable")
	flag.StringVarP(&host,			'j', "host",			"localhost", 	"server host where we're running")
	flag.StringVarP(&serverPort,	'p', "port", 			":3554", 		"port where StreamDude server is listening")
	flag.StringVarP(&frontEnd,		'f', "frontend", 		"nginx", 		"type of frontend/reverse proxy")
	flag.StringVarP(&externalPort,	'P', "externalport",	":80",			"external port if using a reverse proxy")
	flag.StringVarP(&externalHost,	'x', "externalhost",	hostname,		"external hostname if using a reverse proxy")
	flag.StringVarP(&templatePath,	't', "templatepath",	"./templates",	"where templates are held")
	flag.StringVarP(&pathToStaticFiles, 's', "staticpath",	"./assets",		"where static assets are stored")
	flag.StringVarP(&urlPathPrefix,	'u', "urlprefix",		"",				"URL path prefix")
	flag.StringVarP(&lslSignaturePIN, 'l',	"lslpin",		"0000",			"LSL signature PIN")
	flag.BoolVarP(&debug,			'd', "debug",			false, 			"set debug level (omit for normal logs)")
	flag.StringVarP(&streamerURL,	'r', "streamer",		"rtsp://127.0.0.1:554/",	"streamer URL")
	flag.StringVarP(&lalMasterKey,	'k', "masterkey",		"",				"lal server master key")

	flag.Parse()

	if help {
		flag.Usage()
		os.Exit(0)
	}

	// setup a single instance of the validator service.
	validate = validator.New()
	/**
	 * Starting backend web server using Gin Gonic.
	 */
	router := gin.Default()
	router.Delims("{{", "}}") // stick to default delims for Go templates.
	// router.SetTrustedProxies(nil)	// as per https://pkg.go.dev/github.com/gin-gonic/gin#readme-don-t-trust-all-proxies (gwyneth 20220111).
	router.SetTrustedProxies([]string{"127.0.0.1"})	// apparently we should at least trust "our" proxy
	router.TrustedPlatform = gin.PlatformCloudflare	// we're running behind Cloudflare CDN, this will retrieve the correct IP address. Hopefully.

	// Configure logrus.
	//	log.SetFlags(log.LstdFlags | log.Lshortfile)

	logme.Formatter = new(logrus.TextFormatter)
	logme.Formatter.(*logrus.TextFormatter).ForceColors = activeSystemd	// if systemd is active, force colours on log
	logme.Formatter.(*logrus.TextFormatter).DisableColors = false		// keep colors
	logme.Formatter.(*logrus.TextFormatter).DisableTimestamp = false	// keep timestamp

	// set debug level, depending on the argument value
	if (debug) {
		logme.SetLevel(logrus.DebugLevel)
	}

	// respect CLICOLOR_FORCE and NO_COLOR in Gin (logrus is already compliant)
	// Figure out if we're running in a terminal, and, if so, apply all the relevant commands
	// See https://bixense.com/clicolors/ and https://no-color.org/ (gwyneth 20230731)
	// Note: when we're sending logs via journald, we force coloured output, because
	// journald supports it even if it's not a TTY.

	var isTerm = true	// are we logging to a tty?

	logme.Debugf("terminal type: %q activeSystemd: %t NO_COLOR: %q CLICOLOR_FORCE: %q\n",
		os.Getenv("TERM"), activeSystemd, os.Getenv("NO_COLOR"), os.Getenv("CLICOLOR_FORCE"))

	// for the weird type casting, see https://github.com/mattn/go-isatty/issues/80#issuecomment-1470096598 (gwyneth 20230801)
	if (os.Getenv("TERM") == "dumb" || os.Getenv("TERM") == "") && !activeSystemd ||
		(!isatty.IsTerminal(logme.Out.(*os.File).Fd()) && !isatty.IsCygwinTerminal(logme.Out.(*os.File).Fd())) {
			isTerm = false
	}
	if _, ok := os.LookupEnv("NO_COLOR"); ok {
		gin.DisableConsoleColor()
	} else if _, ok := os.LookupEnv("CLICOLOR_FORCE"); (ok && isTerm) || activeSystemd {
		gin.ForceConsoleColor()
	}
	logme.Debugf("debugging to terminal? - %t\n", isTerm)
	// at this stage we have the logging output well configured
	logme.Debugf("Logging debug level set to %q\n", logme.GetLevel().String())


	// Override lal master key from environment.
	if temp := os.Getenv("LAL_MASTER_KEY"); temp != "" {
		lalMasterKey = temp
	}
	if lalMasterKey == "" {
		logme.Warningln("lal master key not found or empty; streaming will probably not work.")
	} else {
		logme.Debugf("lal key (obfuscated): %q\n", obfuscate(lalMasterKey))
	}

	// Override streamer, if env exists.
	if temp := os.Getenv("STREAMER_URL"); temp != "" {
		streamerURL = temp
	}
	// Validate that the streamer has a valid URL (either from command-line or env var).
	if err := validate.Var(streamerURL, "required,url"); err != nil {
		logme.Fatalf("invalid streamer URL: %q, aborting\n", streamerURL)
	}
	logme.Infof("remote streamer URL set to: %q\n", streamerURL)

	if err := validate.Var(externalHost, "hostname_rfc1123,omitempty"); err != nil {
		logme.Errorf("invalid external host name: %q, reverting to empty string\n", externalHost)
		externalHost = ""
	}
	logme.Infof("external hostname set to: %q (empty is ok)\n", externalHost)

	// Setup templating system.

	// var err error	// needed for scope issues
	if workingDirectory, err = os.Getwd(); err != nil {
		workingDirectory = "."	// if os.Getwd() fails, use local directory, maybe it works (gwyneth 2022011.
		// no need to panic, this error is 'fixed'!
	}
	// Figure out where the templates are: deal with empty path.
	if templatePath == "" {
		templatePath = filepath.Join(workingDirectory, "/templates")
	}
	htmlGlobFilePath := filepath.Join(templatePath, "/*.tpl")
	logme.Infof("loading templates from pathToStaticFiles: %q, templatePath: %q, final destination: %q\n",
		pathToStaticFiles, templatePath, htmlGlobFilePath)

	router.LoadHTMLGlob(htmlGlobFilePath)

	// Some useful static dirs & files
	router.Static(path.Join(urlPathPrefix, "/assets"), filepath.Join(pathToStaticFiles, "/assets"))

	router.StaticFile(path.Join(urlPathPrefix, "/favicon.ico"), filepath.Join(pathToStaticFiles, "/assets/favicons/favicon.ico"))
	router.StaticFile(path.Join(urlPathPrefix, "/browserconfig.xml"), filepath.Join(pathToStaticFiles, "/assets/favicons/browserconfig.xml"))
	router.StaticFile(path.Join(urlPathPrefix, "/site.webmanifest"), filepath.Join(pathToStaticFiles, "/assets/favicons/site.webmanifest"))

	// Make the router handle these exceptions with better HTTP error codes
	router.HandleMethodNotAllowed = true
	router.RedirectTrailingSlash = true
	router.RedirectFixedPath = true

	// Generic funcionality

	// Ping handler (who knows, it might be useful in some contexts... such as Let's Encrypt certificates
	router.Any(path.Join(urlPathPrefix, "/ping"),			uiPing)

	// Main website, as far as we can call it a "website".
	router.GET(path.Join(urlPathPrefix, "/home"), 			homepage)
	router.GET(urlPathPrefix + string(os.PathSeparator),	homepage)

	// Shows the credits page.
	router.GET(path.Join(urlPathPrefix, "/credits"),		uiCredits)

	// Lower-leval API for calling things (mostly non-tty low-level calls)
	apiRoutes := router.Group(path.Join(urlPathPrefix, "/api"))
	{		// base page for complex scripts.
		apiRoutes.POST("/play",	apiStreamFile)
		apiRoutes.POST("/auth",	apiSimpleAuthGenKey)
		apiRoutes.POST("/delete", apiDeleteToken)
	}

	// Specific routes just for the user interface
	uiRoutes := router.Group(path.Join(urlPathPrefix, "/ui"))
	{
		uiRoutes.GET("/auth", func(c *gin.Context) {
			// not much to pass really
			c.HTML(http.StatusOK, "form-auth.tpl", environment(c, gin.H{
			}))
		})
		uiRoutes.GET("/play", func(c *gin.Context) {
			// not much to pass really
			c.HTML(http.StatusOK, "form-play.tpl", environment(c, gin.H{
			}))
		})

	}

	// Catch all other routes and send back an error
	router.NoRoute(func(c *gin.Context) {
		errorMessage := "Command " + c.Request.URL.Path + " not found."
		checkErrReply(c, http.StatusNotFound, errorMessage, fmt.Errorf("(routing error)"))
	})

	router.NoMethod(func(c *gin.Context) {
		errorMessage := "Method " + c.Request.Method + " not allowed"
		checkErrReply(c, http.StatusMethodNotAllowed, errorMessage, fmt.Errorf("(unsupported method)"))
	})

	/*
	 *  Deal with Unix system signals (at least those we can catch)
	 */

	// prepares a special (buffered) channel to look for termination signals.
	sigs := make(chan os.Signal, 1)
	// signal.Notify(sigs)	// Note: this should catch all catchable signals!
	signal.Notify(sigs, syscall.SIGHUP, syscall.SIGUSR1, syscall.SIGUSR2, syscall.SIGCONT)

	// goroutine which listens to signals
	// Will handle re-configurations in the future.
	// For now, it exists mostly to signal systemd
	go func() {
		for {
			sig := <-sigs
			switch sig {
				case syscall.SIGUSR1:
					daemon.SdNotify(false, daemon.SdNotifyReloading + "\nSTATUS=reloading request received, but config reload is not supported yet")
					logme.Infoln("SIGUSR1 received, ignoring; we might reload config one day")
					daemon.SdNotify(false, daemon.SdNotifyReady)
				case syscall.SIGUSR2:
					logme.Infoln("SIGUSR2 received, ignoring")
				case syscall.SIGHUP:
					// Note: we *might* interpret this to suspend the processing and/or reload config (gwyneth 202230804)
					logme.Infoln("SIGHUP received (possibly from systemd): hanging up!")
					// if we were called by systemd, then notify it that we're done.
					// if not, just exit normally.
					daemon.SdNotify(true, daemon.SdNotifyStopping)
					os.Exit(129)
				case syscall.SIGCONT:
					logme.Infoln("SIGCONT received, ignoring")
				default:
					// should never happen...?
					logme.Warning("Unknown UNIX signal", sig, "caught, ignoring")
			}
		}
	}()

	// attempt to talk to systemd to notify we're now ready
	b, err = daemon.SdNotify(false, daemon.SdNotifyReady)
	switch {
		case !b && err == nil:
			// the logging system is not available, either, so we just print out
			logme.Warningln("systemd not available")
			activeSystemd = false
		case !b && err != nil:
			logme.Warningln("systemd answered with error:", err)
		case b && err == nil:
			logme.Infoln("systemd was succesfully notified that we're ready")
		default:
			logme.Warningln("unknown/confused systemd status, ignoring")
	}

	/*
	 *  Launch the server (finally) and log an error if it crashes.
	 */

	// this might require another layer to check for https
	errGin := router.Run(host + serverPort)

	// Notify systemd that we're peacefully stopping
	b, err = daemon.SdNotify(true, daemon.SdNotifyStopping  + "\nEXIT_STATUS=126")
	switch {
		case !b && err == nil:
			// the logging system is not available, either, so we just print out
			logme.Warningln("systemd not available")
			activeSystemd = false
		case !b && err != nil:
			logme.Warningln("systemd answered with error:", err)
		case b && err == nil:
			logme.Infoln("systemd was succesfully notified that we're stopping")
		default:
			logme.Warningln("unknown/confused systemd status, ignoring")
	}
	if errGin != nil {
		logme.Errorln("Gin aborted with", errGin)
	} else {
		logme.Errorln("Unexpected error, Gin terminated abruptly without error code")
	}
	os.Exit(126)
}